### ISSUE TITLE
Edge case: `mypy` ignores `Never` being passed as an argument to another function.

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#15821]
**Bug Report**

This is kind of a version 2.0 for https://github.com/python/mypy/issues/15818, as I found an edge-case where `mypy` silently passes an instance of `Never`.

```python
from typing_extensions import Never

def never_returns() -> Never:
    raise NotImplementedError

def g(x: str) -> str:
    return x

def h() -> str:
    return g(never_returns())  # no warning raised here

g(never_returns())  # no warning raised here
# only if there was code after it we would get [unreachable]-error.
```

([mypy-play](https://mypy-play.net/?mypy=master&python=3.11&flags=strict%2Cwarn-unreachable&gist=381e40256253354cd2845044c324110f) with `--strict` and `--warn-unreachable`)

**Expected Behavior**

The above code should raise on lines containing `g(never_returns())`, as instances of `Never` cannot exist, and hence cannot be passed as arguments to other functions. Therefore, the only safe behavior is that functions that return `Never` can only appear "on their own", their output cannot be assigned to variables (the same code as above with `x=never_returns()` in the last line gives the unhelpful message "_Need type annotation for "x"_") or be part of an expression:

```python
never_returns()   # OK
(never_returns() for x in range(3))  # maybe OK?
x = never_returns() # not OK
g(never_returns()) # not OK
```

The `--warn-unreachable` flag is useless when the offending code appears in the last line of a script / function definition.

**Comment by erictraut** [comment]
Mypy is correct in not generating an error here. The type `Never` is a subtype of all other types, so it's type compatible with `str`. Pyright also does not generate an error here for the same reason.

Practically speaking, if the function `never_returns` never returns control back to the caller, then it doesn't violate any type rules for the caller.

**Comment by randolf-scholz** [comment]
@erictraut I completely agree that `Never` is compatible with `str`, but the issue is that for it to be able to be passed, an instance of `Never` has had to exist in the first place. But instances of `Never` are not allowed:

Python docs:

> The [bottom type](https://en.wikipedia.org/wiki/Bottom_type), **a type that has no members.**

Cited Wikipedia article:

> If a type system is [sound](https://en.wikipedia.org/wiki/Type_safety), the bottom type is uninhabited and a term of bottom type represents a logical contradiction. In such systems, typically no distinction is drawn between the bottom type and the [empty type](https://en.wikipedia.org/wiki/Empty_type), and the terms may be used interchangeably.
> 
> If the bottom type is inhabited, its terms[s] typically correspond to error conditions such as undefined behavior, infinite recursion, or unrecoverable errors.

If a type can have no members, a type-checker should probably raise an error if such a member was provably brought into existence.

**Comment by randolf-scholz** [comment]
> Practically speaking, if the function never_returns never returns control back to the caller, then it doesn't violate any type rules for the caller.

But even under this view, shouldn't it at least produce an unreachable-error? Since the calling of the outer function is never reached?

**Comment by erictraut** [comment]
I don't think it's a bug, so I would categorize it as an enhancement request. The maintainers of mypy can decide what they would like to do here. As for pyright, I don't plan to emit any errors here because it's not a type violation and it's not a common bug pattern.

**Comment by randolf-scholz** [comment]
So the second half sentence in _"The [bottom type](https://en.wikipedia.org/wiki/Bottom_type), a type that has no members."_ is basically meaningless to type-checkers?

**Comment by JelleZijlstra** [comment]
The fact that a type has no members doesn't imply that it is an error to use it.

**Comment by randolf-scholz** [comment]
> The fact that a type has no members doesn't imply that it is an error to use it.

It's a logical contradiction to simultaneously assume (1) instances of `Never` do not exist and (2) an instance of `Never` was passed as an argument to another function.

**Comment by A5rocks** [comment]
To chime in with my opinion:

`Never` doesn't exist as far as type checkers know. However, being able to pass it places has (very!) practical benefits such as:
 - making sure exhaustive checks are exhaustive
 - deprecations that aren't in the type signature but you still warn about
 - more helpful handling of type-level impossible values

We're better off for all of those.

As such, practicality over purity and this is allowed. At least that's how it seems to me. If this is actually causing you errors, maybe things can be rethought.

**Comment by randolf-scholz** [comment]
@A5rocks Can you give an example?

> making sure exhaustive checks are exhaustive

Not sure how the proposal here would stop you from doing exhaustive checks.

> deprecations that aren't in the type signature but you still warn about

We'll have https://peps.python.org/pep-0702/ for that.

> more helpful handling of type-level impossible values

When and how is this useful?

**Comment by A5rocks** [comment]
Here's an example of exhaustiveness checking, with a deprecation:

```py
import enum
import typing

class A(enum.Enum):
  x = "X"
  y = "Y"

def f(x: A) -> str:
  if not isinstance(x, A):
    # this function used to take "X" or "Y", maybe?
    print(f"Make sure you pass `A.x` or `A.y`; not {x!r}")
    x = A(x)  # we're using a Never here

  if x is A.x:
    return "A"
  elif x is A.y:
    return "B"
  else:
    typing.assert_never(x)  # we're using another Never here
```

**Comment by randolf-scholz** [comment]
`typing.assert_never` would unsurprisingly need to be special-cased, this is a bit of an alibi example.

For the first part, `f` should be annotated as `f(x: A | Any)` or have overloads if it is supposed to accept things different from `A`.

```python
@overload
def f(x: A) -> str: ...
@overload
@deprecated("Only A will be allowed in the future")
def f(x: Other) -> str: ...
```

**Comment by JelleZijlstra** [comment]
`assert_never()` isn't meant to be special-cased. Once again, what you're looking for is not `Never` but some new primitive. As others have explained, there's no bug here.

**Comment by randolf-scholz** [comment]
Then the docs need to be changed, as

> The [bottom type](https://en.wikipedia.org/wiki/Bottom_type), a type that has no members.

Clearly is violated by the major type-checkers.

**Comment by JelleZijlstra** [comment]
We can discuss a documentation change over on the CPython repo, but I wrote that documentation and I don't think "has no members" implies that type checkers should error on usage of Never.

**Comment by randolf-scholz** [comment]
@JelleZijlstra The following doesn't emit any errors, even with `--strict`. 

```python
from typing import Never

def f() -> Never:
    while True:
        pass
    
x: Never = f()
```

So the type checker has allowed me to create a member of `Never`.
